### PR TITLE
Issue 874 : add custom events for Metrics

### LINKED ISSUE #874: Support custom events for metrics update
Currently for ignite metrics, `update` function is called per iteration. In scenarios like multitasking, there can be metrics for different tasks, and each iteration might only have outputs for a subset of tasks. In this situation, only the metrics associated with those tasks having outputs can/should be updated.

To make it work, I need to overwrite the `attach` function like this:
```
class CustomRnningAverage(RunningAverage):
    def attach(self, engine, name, custom_event=None):
        if custom_event is None:
            super().attach(engine, name)
        else:
            if self.epoch_bound:
                # restart average every epoch
                engine.add_event_handler(Events.EPOCH_STARTED, self.started)
            # compute metric
            engine.add_event_handler(custom_event, self.iteration_completed)
            # apply running average
            engine.add_event_handler(custom_event, self.completed, name)
```
This is doable, but would be nice to have it by default in ignite. Also, this manual overwrite doesn't work with DDP. The cause is related to `_internal_attach` in `metrics_lambda.py`.

I would like to propose
1.  Have `update` every iteration by default
2. Support overwrite when to `update` by accepting custom events
3. Make it work with DDP

Multitasking is one case where this can be very helpful. But I am sure there will be other use cases.

### PR DESCRIPTION
**Pull request by sdesrozis**
Fixes #874 

Description:

Allow customization for `Events` of `Metrics`.

Check list:
* [x] New tests are added (if a new feature is added)
* [x] New doc strings: description and/or example code are in RST format
* [ ] Documentation is updated (if required)

### CODE REVIEW

**Comment by sdesrozis** [comment]
@vfdev-5 thoughts on this draft ?

**Comment by vfdev-5** [comment]
We may have to take this comment into account for the usage : https://github.com/pytorch/ignite/pull/644#issuecomment-576557938

**Comment by sdesrozis** [comment]
> We may have to take this comment into account for the usage : https://github.com/pytorch/ignite/pull/644#issuecomment-576557938

Same issue??

**Comment by sdesrozis** [comment]
@vfdev-5 

Major refactoring
- Add interface `EngineMetric` which defines methods used by `Engine`
- Add functions `attach`, `detach` and `is_attached` external to `Metric` using `Engine`, `MetricUsage` and `EngineMetric`

The idea is to have `Metric` like an algorithm pluggable in any `Engine` wrt a `MetricUsage`. Today, `attach`, `detach` and `is_attached` are embedded in `Metric` but one could use external functions.

Thoughts ?

**Comment by vfdev-5** [comment]
@sdesrozis please, show an example of usage. For instance, I do not understand the need of another interface... Thanks

**Comment by sdesrozis** [comment]
The idea is to have `Metric` like an algorithm with API to be plugged to engine but no reason to have the plug in the metric. 

This is a separation of concerns.

**Comment by vfdev-5** [comment]
@sdesrozis I'm not fan of "possible API", let's keep current API. 
If I get it right, `EngineMetric` interface is only for "possible API" ?

**Comment by sdesrozis** [comment]
> @sdesrozis I'm not fan of "possible API", let's keep current API. 
> 
> If I get it right, `EngineMetric` interface is only for "possible API" ?
> 
> 

Yes, I can remove it, no problem 😊

**Comment by sdesrozis** [comment]
Removed

**Comment by vfdev-5** [comment]
@sdesrozis thanks ! What do you think about to enhance the following API by 
```python
engine1 = ...
engine2 = ...

m = MyMetric()

# current API
m.attach(engine1, "m")
m.attach(engine2, "m", usage="batch_wise")  # ~ BatchWise()
m.attach(engine2, "m", usage="epoch_wise")  # ~ EpochWise()
```
? 

No need to import those classes.

**Comment by sdesrozis** [comment]
Note that `BatchFiltered` needs an argument.

**Comment by sdesrozis** [comment]
Done !

Adding `MetricUsage` in `attach` rather than `__init__` implies that this feature is now usable by all metrics without any modification.

**Comment by vfdev-5** [comment]
@snie2012 could you please take a look and give us your feedback if this PR covers your need in #874, thanks !

@VinhLoiIT could you also take a look on this PR if it covers your use-case to compute batch-wise metrics as mentioned in #848 ?

**Comment by erip** [comment]
> Note that `BatchFiltered` needs an argument.

I'm not sure of a clean way to do this. There are three options for this:

1. A user provides a constructed `BatchFiltered` via `usage`.
1. A user provides `usage="batch_filter` and a kwarg predicate callable.
1. A user provides `usage=batch_filter` and not a kwarg predicate callable.

The logic for validating this is a bit hairy, I think. It would be something like...

```python
def attach(..., usage=..., filter: Optional[Callable] = None):
    if isinstance(usage, BatchFilter) and filter is not None:    
       raise ValueError("Can't provide a BatchFilter and a filter function.")
    elif usage == "batch_filter" and filter is None:
        raise ValueError("A filter function must be provided with \"batch_filter\".")
    elif usage == "batch_filter" and filter is not None:
        batch_filter = BatchFilter(filter)
    else:
        # handle metrics that don't require arguments here
        ... 
```

As more `MetricUsage`s are introduced that require arguments, this might become too messy. Maybe as a rule we say "if your `MetricUsage` requires arguments, its your responsibility to construct it and pass it in" -- WDYT @sdesrozis?

**Comment by sdesrozis** [comment]
I prefer when things are simple 😊 I agree the last part, user is responsible if he does non standard usage (filtered or custom, because he can!) 😉

**Comment by erip** [comment]
> I prefer when things are simple 😊 I agree the last part, user is responsible if he does non standard usage (filtered or custom, because he can!) 😉

I like that. It also doesn't open up `attach` more than necessary.

**Comment by sdesrozis** [comment]
If needed, we can have `detach` and `is_attached` without usage by 1. testing usages (but what happen for custom usages) or removing from engine whatever the usage registered (carful for lambda...).

**Comment by snie2012** [comment]
> @snie2012 could you please take a look and give us your feedback if this PR covers your need in #874, thanks !
> 
> @VinhLoiIT could you also take a look on this PR if it covers your use-case to compute batch-wise metrics as mentioned in #848 ?

Replied in the #874

**Comment by VinhLoiIT** [comment]
Nice work!
I have two test cases mentioned in #848 to be tested:

1. The first one is that I would like to update my metric every batch so that it could be computed and logged with `ProgressBar` during evaluation. Your work solved peacefully by creating the `MetricUsage` instance:
```python
usage = MetricUsage(
    Events.EPOCH_STARTED,
    Events.ITERATION_COMPLETED,
    Events.ITERATION_COMPLETED,
)

MyMetric().attach(engine, 'MyMetric', usage=usage)
```

2. The second one is that I would like to reset my metric during training (similar to parameter `log_interval` in pure PyTorch print-style example. This case is useful in terms of calculating `Loss` (which is basically inherited from `Metric`) that it should be reset every `log_interval` while training to not be affected by averaging in the long run. I think it would better to use `RunningAverage` from ignite, but I have to add an event to reset by myself. To solve this by `MetricUsage`, it is equivalent to:
```python
usage = MetricUsage(
    Events.ITERATION_STARTED(every=log_interval), # started
    Events.ITERATION_COMPLETED,                   # completed
    Events.ITERATION_COMPLETED,                   # iteration_completed
)
```
However, it does not work as expected due to the exception:
```python
  ...
  File "/home/vinhloi/projects/fork/ignite/metrics/metric.py", line 278, in attach
    if not engine.has_event_handler(self.started, usage.STARTED):
  File "/home/vinhloi/projects/fork/ignite/engine/engine.py", line 303, in has_event_handler
    self._assert_non_filtered_event(event_name)
  File "/home/vinhloi/projects/fork/ignite/engine/engine.py", line 291, in _assert_non_filtered_event
    "Argument event_name should not be a filtered event, " "please use event without any event filtering"
TypeError: Argument event_name should not be a filtered event, please use event without any event filtering
```
I just wonder how to solve this problem. The `MetricUsage` as above logically true, I guess

**Comment by sdesrozis** [comment]
@VinhLoiIT Thank for the report :)

For point 2. it's out of the scope of `MetricUsage` at the moment. There are too much scenarii of use and it's very easy to write bugs. The supported usages are `EpochWise`, `BatchWise` and `BatchFiltered`. We will warn about that and improve the document to be extremely clear about the scope. 

Btw, I will check your code to see if I can fix ;)

**Comment by sdesrozis** [comment]
@vfdev-5 

filter can't apply to events of metrics.

See here https://github.com/pytorch/ignite/blob/c3c1ff22343d4f76979595ff7ac0707b49fc7c98/ignite/engine/engine.py#L285

**Comment by sdesrozis** [comment]
@VinhLoiIT 

At the moment, we can't use filtered event with `MetricUsage`. However, it works with custom events. A workaround (ugly) could be
```python
# custom event for Events.ITERATION_COMPLETED(every=log_interval)
class MyEvent(EventEnum):
    ITERATION_STARTED = 'MyEvent_ITERATION_STARTED'

engine.register_events(*MyEvent)

usage = MetricUsage(
    started=MyEvent.ITERATION_STARTED, # use custom event
    completed=Events.ITERATION_COMPLETED,
    iteration_completed=Events.ITERATION_COMPLETED,
)

# to apply filter every log_interval
@engine.on(Events.ITERATION_COMPLETED(every=log_interval))
def _():
    engine.fire_event(MyEvent.IT_STARTED)
```

I can't do better :(

**Comment by vfdev-5** [comment]
> @vfdev-5
> 
> filter can't apply to events of metrics.
> 
> See here
> 
> https://github.com/pytorch/ignite/blob/c3c1ff22343d4f76979595ff7ac0707b49fc7c98/ignite/engine/engine.py#L285

@sdesrozis can't we relax this condition ?

**Comment by sdesrozis** [comment]
> @sdesrozis can't we relax this condition ?

We can do it but I don't have history about why this condition is enabled.

**Comment by vfdev-5** [comment]
> > @sdesrozis can't we relax this condition ?
> 
> We can do it but I don't have history about why this condition is enabled.

This `_assert_non_filtered_event` was added before unification of `Events` into `EventsEnum` etc. Now, we must ensure that `has_event_handler` works correctly for filtered events, do not pick a wrong handlers instead etc and it can be enough.

**Comment by sdesrozis** [comment]
I finish this PR this week-end. Sorry for the lag.

**Comment by sdesrozis** [comment]
@vfdev-5 what do you think ?

File: ignite/engine/engine.py [hunk 1]
```diff
@@ -300,7 +300,7 @@ def has_event_handler(self, handler: Callable, event_name: Optional[Any] = None)
                 to ``None`` to search all events.
         """
         if event_name is not None:
-            self._assert_non_filtered_event(event_name)
+            # self._assert_non_filtered_event(event_name)
```
Comment (vfdev-5) [inline ignite/engine/engine.py:303]: We can remove commented code. Have you checked the correct behaviour of `has_event_handler` on filtered events ?

File: ignite/engine/engine.py [hunk 2]
```diff
@@ -327,7 +327,7 @@ def remove_event_handler(self, handler: Callable, event_name: Any):
             event_name: The event the handler attached to.
 
         """
-        self._assert_non_filtered_event(event_name)
+        # self._assert_non_filtered_event(event_name)
```
Comment (vfdev-5) [inline ignite/engine/engine.py:330]: Same here: remove commented code and if not yet done, we need a teston checking the correct behaviour with filtered events

File: ignite/engine/engine.py (hunk 1 above)
Comment (sdesrozis) [reply ignite/engine/engine.py:303]: It seems correct. However I noticed that if a handler is registered with a filtered event, you can remove it using non filtered event (Ser test_custom_event.py)

**Comment by sdesrozis** [comment]
Usage `epoch_wise` is well tested. Need use cases for others usages.

**Comment by vfdev-5** [comment]
@sdesrozis code looks OK, documentation layout is not correct:
- missing docs for "MetricUsage", "EpochWise", "BatchWise", "BatchFiltered"
- missing indents renders badly docs, i'll comment out where.
Please, try to render the docs locally and check it.

File: ignite/metrics/metric.py [hunk 3]
```diff
@@ -10,7 +10,86 @@
...
+    def ITERATION_COMPLETED(self):
+        return self.__iteration_completed
+
+
+class EpochWise(MetricUsage):
+    """
+    Epoch-wise usage of Metrics. It's the default and most common usage of metrics.
+
+    Metrics :
+    * starts to compute at Events.EPOCH_STARTED,
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:46]: I'm not sure if this would render correctly. To check.

File: ignite/metrics/metric.py [hunk 4]
```diff
@@ -147,14 +232,28 @@ def completed(self, engine: Engine, name: str) -> None:
...
+    def attach(self, engine: Engine, name: str, usage: Union[str, MetricUsage] = EpochWise()) -> None:
         """
         Attaches current metric to provided engine. On the end of engine's run,
         `engine.state.metrics` dictionary will contain computed metric's value under provided name.
 
         Args:
             engine (Engine): the engine to which the metric must be attached
             name (str): the name of the metric to attach
+            usage (str or MetricUsage, optional): the usage of the metric. Valid string values should be
+            'epoch_wise' (default) or 'batch_wise'.
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:256]: Missing indent here

File: ignite/metrics/metric.py [hunk 5]
```diff
@@ -182,6 +293,8 @@ def detach(self, engine: Engine) -> None:
 
         Args:
             engine (Engine): the engine from which the metric must be detached
+            usage (str or MetricUsage, optional): the usage of the metric. Valid string values should be
+            'epoch_wise' (default) or 'batch_wise'.
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:297]: Missing indent here

File: ignite/metrics/metric.py [hunk 6]
```diff
@@ -194,23 +307,39 @@ def detach(self, engine: Engine) -> None:
...
+
+    def is_attached(self, engine: Engine, usage: Union[str, MetricUsage] = EpochWise()) -> bool:
         """
         Checks if current metric is attached to provided engine. If attached, metric's computed
         value is written to `engine.state.metrics` dictionary.
 
         Args:
             engine (Engine): the engine checked from which the metric should be attached
+            usage (str or MetricUsage, optional): the usage of the metric. Valid string values should be
+            'epoch_wise' (default) or 'batch_wise'.
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:339]: Missing indent here

**Comment by sdesrozis** [comment]
@vfdev-5 thak you, I update soon :p

**Comment by sdesrozis** [comment]
Doc done. Next, I would like to refactor a last thing after merge of fix for #1004 (if ok).

**Pull request merged**